zanata-client-commands/src/main/java/org/zanata/client/commands/push/RawPushCommand.java

```diff
@@ -75,10 +94,164 @@ public RawPushCommand(PushOptions opts, RestClientFactory clientFactory) {
+                continue;
+            }
+
+            //throw error if file type is input more than once
+            if (filteredFileTypes.containsKey(inputFileType)) {
+                String msg =
+                    "File type '" + fileType + "' cannot be used more than once in option --file-types.";
+                log.error(msg);
+                throw new RuntimeException(msg);
+            }
+
+            Set<String> filteredExtensions = new HashSet<>(extensions);
+
+            if (!extensions.isEmpty()) {
+                //throw exception if user input extension if not part of DocumentType supported source extension
+                for (String extension: extensions) {
+                    if (!inputFileType.getSourceExtensions()
+                            .contains(extension)) {
+                        String msg =
+                                "Extension '"
+                                        + extension
+                                        + "' for file type '"
+                                        + fileType
+                                        + "' is not supported by the target server and will be ignored.";
+                        consoleInteractor.printfln(DisplayMode.Warning, msg);
+                        filteredExtensions.remove(extension);
+                    }
+                }
+            }
+
+            /**
+             * Use the extensions from typeWithExtension input if exists,
+             * otherwise, use the extensions from server.
+             */
+            filteredExtensions =
+                filteredExtensions.isEmpty() ? inputFileType.getSourceExtensions()
+                            : filteredExtensions;
+
+            //throw error if same file extension found in multiple file type
+            for (DocumentType filteredDocType: filteredFileTypes.keySet()) {
+                for (String filteredExtension: filteredDocType.getSourceExtensions()) {
+                    if (filteredExtensions.contains(filteredExtension)) {
+                        String msg =
+                                "Conflict file type between '" + fileType
+                                        + "' and '" + filteredDocType.name() + "' "
+                                        + "for extension '"
+                                        + filteredExtension
+                                        + "' in option --file-types.";
+                        log.error(msg);
+                        throw new RuntimeException(msg);
+                    }
+                }
+            }
+            filteredFileTypes.put(inputFileType, filteredExtensions);
+        }
+        return filteredFileTypes;
+    }
+
     @Override
     public void run() throws IOException {
         PushCommand.logOptions(log, getOpts());
-        log.warn("Using EXPERIMENTAL project type 'file'.");
+        consoleInteractor.printfln(DisplayMode.Warning,
+            "Using EXPERIMENTAL project type 'file'.");
```

Comment:
- [ ] @aeng it looks like you stopped calling it experimental. Do our QA/UX guys think it is stable and feature-complete enough for that?